### ISSUE TITLE
Deprecate RSA_SSLV23_PADDING 

### ISSUE THREAD

**Issue by kroeckx** [openssl/openssl#14216]
With the CVE-2021-23839 change, there is at least 1 test suite that starts to fail where it does an encrypt using that padding mode, and then decrypts using the same padding mode. I think that's because of a misunderstanding of how the padding mode is supposed to work, and the error is really what should have happened.

My understanding of it that if you're talking SSLv2, but SSLv3 is supported, you should set that mode. This has as effect that if both sides of a connection use that padding mode, and so support SSLv3, but are talking SSLv2 the connection should fail. If the other doesn't support SSLv3, it would have used the RSA_PKCS1_PADDING padding mode and would not get an error.

**Comment by mattcaswell** [comment]
> My understanding of it that if you're talking SSLv2, but SSLv3 is supported, you should set that mode. This has as effect that if both sides of a connection use that padding mode, and so support SSLv3, but are talking SSLv2 the connection should fail. If the other doesn't support SSLv3, it would have used the RSA_PKCS1_PADDING padding mode and would not get an error.

Your understanding is correct

**Comment by kaduk** [comment]
I think the pithy answer to "how to use `RSA_SSLV23_PADDING`" is just "don't", to go along with "don't use SSLv2".

**Comment by kroeckx** [comment]
I guess we should deprecate that padding mode.

**Comment by mcepl** [comment]
> I think the pithy answer to "how to use `RSA_SSLV23_PADDING`" is just "don't", to go along with "don't use SSLv2".

Well, my understanding (speaking as the upstream maintainer of M2Crypto, which is the package in question) was that “`sslv23`” has for long time nothing to do with SSLv2 and it means just “do the best thing possible”. So, the code in question:

```python

self = <tests.test_rsa.RSATestCase testMethod=test_public_encrypt>

    @unittest.skipIf(m2.OPENSSL_VERSION_NUMBER < 0x1010103f,
                     'Relies on fix which happened only in OpenSSL 1.1.1c')
    def test_public_encrypt(self):
        priv = RSA.load_key(self.privkey)
        # pkcs1_padding, pkcs1_oaep_padding
        for padding in self.e_padding_ok:
            p = getattr(RSA, padding)
            ctxt = priv.public_encrypt(self.data, p)
            ptxt = priv.private_decrypt(ctxt, p)
            self.assertEqual(ptxt, self.data)

        # sslv23_padding
        ctxt = priv.public_encrypt(self.data, RSA.sslv23_padding)
        res = priv.private_decrypt(ctxt, RSA.sslv23_padding)
```
does nothing wrong. Does it?

**Comment by davidben** [comment]
`SSLv23_method`, `SSLv23_client_method`, and `SSLv23_server_method`, in libssl did mean to do SSL/TLS with version negotiation. This was very confusing so it's now `TLS_method`, etc.

`RSA_SSLV23_PADDING` is something else entirely and should not be used. Parameter negotiation happens at much much higher level than the internals of the padding, so it wouldn't make sense to "do the best thing possible" here.

**Comment by mcepl** [comment]
So, what should I do?

**Comment by davidben** [comment]
What are you trying to do? This padding mode only makes sense as part of an SSLv2 implementation. Unless your goal is to implement SSLv2, don't use it.

**Comment by mcepl** [comment]
@davidben It should be just simple `encrypted-decrypted-text == original-plaintext` with different parameters. I have fixed the code above so it shows whole test case.

**Comment by t8m** [comment]
@mcepl Yes, the simplest and most sane thing is to just delete the testcase with the sslv23_padding.

**Comment by t8m** [comment]
I've transformed this issue into deprecation issue for the RSA_SSLV23_PADDING.

**Comment by davidben** [comment]
@mcepl I would suggest deleting, or at least deprecating, the bindings too. I doubt anyone is using M2Crypto to implement SSL 2.0 in Python. :-)

**Comment by mcepl** [comment]
> @mcepl I doubt anyone is using M2Crypto to implement SSL 2.0 in Python. :-)

I will wait on you before doing it. Considering total depravity of humankind, I am certain that somebody uses SSL 2.0, hopefully just for internal purposes. And if you wouldn’t believe in the universal tendency of humankind towards sin just look how we deal with security on Internet, it must persuade you.

No, seriously, main purpose of M2Crypto is to support legacy applications based on it, new application in my opinion should use python-cryptography, and so I try to be as conservative as possible, or even more.

**Comment by t8m** [comment]
@mcepl there is an ongoing OTC vote and it seems we are going to remove RSA_SSLV23_PADDING completely from OpenSSL 3.0.

**Comment by t8m** [comment]
The vote has passed. I am closing this issue as we have another issue #14283 for the removal.

**Comment by jeff-h** [comment]
Not sure if it's helpful (or possibly obvious)... but I just checked my previous builds (from around April) and all the PHP versions (right up to and including 8.3) have a dependency on `libssl.1.1.dylib` but the ones I *can* build successfully now depend only on `libssl.3.dylib`.
